modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/QueryCursorSpliteratorCallsTest.java

```diff
@@ -0,0 +1,117 @@
+import org.apache.ignite.cache.query.TextQuery;
+import org.apache.ignite.configuration.CacheConfiguration;
+import org.apache.ignite.configuration.IgniteConfiguration;
+import org.apache.ignite.testframework.GridTestUtils;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.junit.Test;
+
+/**
+ * Query cursor spliterator called multiple times without triggering IgniteException("Iterator is already fetched or
+ * query was cancelled.")
+ */
+public class QueryCursorSpliteratorCallsTest extends GridCommonAbstractTest {
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        return super.getConfiguration(igniteInstanceName)
+            .setCacheConfiguration(new CacheConfiguration<>(DEFAULT_CACHE_NAME)
+                .setAtomicityMode(CacheAtomicityMode.TRANSACTIONAL)
+                .setIndexedTypes(Integer.class, String.class));
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTestsStarted() throws Exception {
+        startGrids(1);
+    }
+
+    /**
+     * @throws IgniteException If failed.
+     */
+    @Test
+    public void testScanQueryCursorSpliteratorCalls() throws IgniteException {
+        doQueryCursorSpliteratorCalls(new ScanQuery<>((key, val) -> key != null));
+    }
+
+    /**
+     * @throws IgniteException If failed.
+     */
+    @Test
+    public void testContinuousQueryCursorSpliteratorCalls() throws IgniteException {
+        doQueryCursorSpliteratorCalls(new ContinuousQuery<>()
+            .setInitialQuery(new ScanQuery<>((key, val) -> key != null))
+            .setAutoUnsubscribe(true)
+            .setLocalListener(iterable -> {}));
+    }
+
+    /**
+     * @throws IgniteException If failed.
+     */
+    @Test
+    public void testSqlQueryCursorSpliteratorCalls() throws IgniteException {
+        doQueryCursorSpliteratorCalls(new SqlQuery<>("String", "from String"));
+    }
+
+    /**
+     * @throws IgniteException If failed.
+     */
+    @Test
+    public void testSqlFieldQueryCursorSpliteratorCalls() throws IgniteException {
+        doQueryCursorSpliteratorCalls(new SqlFieldsQuery("select _key from String"));
+    }
+
+    /**
+     * @throws IgniteException If failed.
+     */
+    @Test
+    public void testTextQueryCursorSpliteratorCalls() throws IgniteException {
+        doQueryCursorSpliteratorCalls(new TextQuery<>("String", "1"));
+    }
+
+    /**
+     * Executes query on cache then calls {@link QueryCursor#iterator()} and {@link QueryCursor#spliterator()}
+     * sequentially.
+     *
+     * @param qry Query.
+     */
+    private void doQueryCursorSpliteratorCalls(Query<?> qry) {
+        Ignite client = grid(0);
+
+        IgniteCache<Object, String> cache = client.getOrCreateCache(new CacheConfiguration<>(DEFAULT_CACHE_NAME));
+
+        try (QueryCursor<?> cur = cache.query(qry)) {
+            cur.iterator();
```

Comment:
```
            Iterator<Object> iter = (Iterator<Object>)cur.iterator();
            Spliterator<Object> spliter = (Spliterator<Object>)cur.spliterator();

            Set<Object> iterData = new HashSet<>();
            Set<Object> spliterData = new HashSet<>();

            while (iter.hasNext())
                iterData.add(iter.next());

            spliter.forEachRemaining(spliterData::add);

            assertEqualsCollections(iterData, spliterData);
```

Reply:
This will not work, since you can't reuse QueryCursorImpl's iterator/spliterator. I think that's why there's `new IgniteException("Iterator is already fetched or query was cancelled.")` on iter() call. Will add AutoClosableCursorSpliterator